rclpy/src/rclpy/names.cpp

```diff
@@ -0,0 +1,289 @@
+}
+
+py::object
+get_validation_error_for_full_topic_name(const char * topic_name)
+{
+  int validation_result;
+  size_t invalid_index;
+  rmw_ret_t ret = rmw_validate_full_topic_name(topic_name, &validation_result, &invalid_index);
+  if (ret == RCL_RET_BAD_ALLOC) {
+    rcl_reset_error();
+    throw std::bad_alloc();
+  } else if (ret != RCL_RET_OK) {
+    throw RCLError("failed to validate name");
+  }
+
+  if (validation_result == RMW_NAMESPACE_VALID) {
+    return py::none();
+  }
+
+  const char * validation_message = rmw_full_topic_name_validation_result_string(validation_result);
+
+  return py::make_tuple(validation_message, invalid_index);
+}
+
+py::object
+get_validation_error_for_namespace(const char * namespace_)
+{
+  int validation_result;
+  size_t invalid_index;
+  rmw_ret_t ret = rmw_validate_namespace(namespace_, &validation_result, &invalid_index);
+  if (ret == RCL_RET_BAD_ALLOC) {
+    rcl_reset_error();
+    throw std::bad_alloc();
+  } else if (ret != RCL_RET_OK) {
+    throw RCLError("failed to validate namespace");
+  }
+
+  if (validation_result == RMW_NAMESPACE_VALID) {
+    return py::none();
+  }
+
+  const char * validation_message = rmw_namespace_validation_result_string(validation_result);
+
+  return py::make_tuple(validation_message, invalid_index);
+}
+
+py::object
+get_validation_error_for_node_name(const char * node_name)
+{
+  int validation_result;
+  size_t invalid_index;
+  rmw_ret_t ret = rmw_validate_node_name(node_name, &validation_result, &invalid_index);
+  if (ret == RCL_RET_BAD_ALLOC) {
+    rcl_reset_error();
+    throw std::bad_alloc();
+  } else if (ret != RCL_RET_OK) {
+    throw RCLError("failed to validate node name");
+  }
+
+  if (validation_result == RMW_NODE_NAME_VALID) {
+    return py::none();
+  }
+
+  const char * validation_message = rmw_node_name_validation_result_string(validation_result);
+
+  return py::make_tuple(validation_message, invalid_index);
+}
+
+std::string
+expand_topic_name(const char * topic, const char * node_name, const char * node_namespace)
+{
+  rcutils_allocator_t rcutils_allocator = rcutils_get_default_allocator();
+  rcutils_string_map_t substitutions_map = rcutils_get_zero_initialized_string_map();
+
+  rcutils_ret_t rcutils_ret = rcutils_string_map_init(&substitutions_map, 0, rcutils_allocator);
+  if (rcutils_ret == RCUTILS_RET_BAD_ALLOC) {
+    rcl_reset_error();
+    throw std::bad_alloc();
+  } else if (rcutils_ret != RCUTILS_RET_OK) {
+    throw RCLError("failed to initialize string map");
+  }
+
+  rcl_ret_t ret = rcl_get_default_topic_name_substitutions(&substitutions_map);
+  if (ret == RCL_RET_BAD_ALLOC) {
+    rcl_reset_error();
+    throw std::bad_alloc();
+  } else if (ret != RCL_RET_OK) {
+    auto exception = RCLError("failed to get default substitutions");
+    // finalize the string map before returning
+    rcutils_ret = rcutils_string_map_fini(&substitutions_map);
+    if (rcutils_ret != RCUTILS_RET_OK) {
+      fprintf(
```

Comment:
@sloretz FYI there's `RCUTILS_SAFE_FWRITE_TO_STDERR` too.

Reply:
Oops, I skimmed over the fact you are formatting. Either disregard or drop the integer formatting and use two safe writes.

Reply:
Uses two calls to `RCUTILS_SAFE_FWRITE_TO_STDERR` in 90f1052